### ISSUE TITLE
Document potential "Missing Required Modifications"

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#826]
We are all keenly aware that there might be security bugs in our code. This is why we spend time reviewing pull requests and plan to get security auditors looking at all the modifications we've made to bitcoin core. However, I feel there's a class of bugs that isn't getting enough of our explicit attention.

I foresee three classes of mistakes:

1. A mistake in the code we've *added* to bitcoin core, which breaks the security of the Zcash system.
2. A mistake in one of our *modifications to* bitcoin core, which breaks the security of some system that already exists in bitcoin (p2p protocol, wallet, consensus), even if users never do any zero-knowledge transactions.
3. Changes that we *were supposed to make*, but didn't, either because we didn't even consider making those changes, or we ran out of time.

Our team is already well aware of bug types (1) and (2). It's (3) that I'm worried about. It's worrying because (3) includes "unknown unknowns." By (3) I mean changes we need to make but aren't even *aware of* right now. For example changing the alert broadcast private keys (#424), or dealing with the block heights that are hard-coded into upstream (#420). If we had forgotten about either of those things, we would have shipped a product that gave someone else access to our alert system, or a product that suddenly dies once a certain block height is reached.

What else are we forgetting? There are almost certainly other changes we *need* to make that we aren't even aware of right now.

I suggest we have a separate audit for type-(3) bugs. Our internal security audit and external security audits are probably going to focus on type-(1) and type-(2) bugs without thinking of type-(3) bugs. We need time set aside to think about type-(3) bugs.

**Comment by defuse** [comment]
Concretely, close this issue by having a meeting to brainstorm possible type-(3) issues.

**Comment by daira** [comment]
Let's do the meeting to brainstorm this well in advance, say within the next couple of weeks.

**Comment by defuse** [comment]
It would be useful to collaborate with upstream developers on this, if possible.

**Comment by defuse** [comment]
We can do the brainstorming asynchronously. I don't think well in meetings anyway.

Here's my brainstorm:

- "Names" in the codebase that refer to external things specific to bitcoin that we should also be duplicating. For example, URLs to the GitHub repo (#884) or to the bitcoin website. Names of IRC channels in documentation, etc.

- Things that create undesired interaction between the bitcoin network and our network, e.g. leads to zcash nodes trying to connect to bitcoin nodes, or leads to zcash nodes' IP addresses being broadcast on the bitcoin network, or things like that.

- Infrastructure that's required for the P2P network to work, which *we* haven't needed because we're connecting directly to our testnet. For example, DNS seeds, the thing where it attempts to connect to .onions.

- Time- or blockheight-based modifications of behavior, e.g. special casing of block numbers when new features were introduced, etc.

- Parameter tuning: Zcash is different than bitcoin, so any tweakable knobs upstream bitcoin has set may need to be set differently for Zcash.

- Stuff that may have been important in 2009 for bootstrapping the original Bitcoin network which has since been stripped out of the code because it's already a huge network and it doesn't need that anymore.

(I plan to edit this comment with more ideas).

**Comment by jcb82** [comment]
It looks like there is a quite comprehensive list going so far. I'm not sure I can venture specific additions since I don't know the codebase well enough. I might suggest the following steps to try to be more systematic:

*An audit of all #defined constants in the source code. This is a primary source of things that might have been forgotten to be updated.
*An audit of all strings in the source code for similar reasons.

**Comment by defuse** [comment]
Here's another example (from main.cpp). The comment will become incorrect once we change the block interval:

```
    // If our best fork is no longer within 72 blocks (+/- 12 hours if no one mines it)
    // of our head, drop it
    if (pindexBestForkTip && chainActive.Height() - pindexBestForkTip->nHeight >= 72)
        pindexBestForkTip = NULL;
``

**Comment by daira** [comment]
Add Zcash-specific safe calls to the calls allowed in safe mode.

https://github.com/zcash/zcash/pull/1337#issuecomment-244599148

**Comment by defuse** [comment]
Notes from our collective brainstorm (which happened a long time ago; I just recently found the notes I took from the meeting):

- Dust -- turn it off?
- Pruning / Joinsplit -- Make sure we don't prune stuff we still need (UTXO are spent but still have pours or something)
- Would be nice to simulate launch + activity for x years into the future.
- Split up research work and recombine in a new meeting (I can't remember what this means).
- Interaction with the Bitcoin (and other coin) network. Make sure whatever we changed the protocol ID thing to isn't also used by one of the other altcoins (this can happen by random chance, so we should check).
- Tor
    - Onion / regular partittioning
        - All traffic was going through including cleartext nodes
        - Trusted tor hidden services
            - hardcoed list of hidden services
        - cleartext<-> tor bridge nodes
- Simulating e.g. 10 years in a week
    - Use real-world BTC data
        - Should all fail up to a certain height
- Backup wallet (non-protocol things)
- Task: go over all RPC commands and upgrade (i.e. insure they meet Bitcoin users' expectations and the documented behavior OR disable them).
- Find existing tools that use the RPC commands and use them (integration test).

**Comment by nathan-at-least** [comment]
I've put this in Beta-2, although I'm wary of not having a clear criterion for closing this ticket. I think we will need a timeout where we simply document this brainstorm. :-(

**Comment by nathan-at-least** [comment]
Let's close this after linking to it from our ``security-warnings.md``.

**Comment by daira** [comment]
Will be closed by #1644.
